tools/nnyeah/nnyeah/AssemblyComparator/ModuleVisitor.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.IO;
+using Mono.Cecil;
+using System.Collections.Generic;
+using System.Linq;
+
+#nullable enable
+
+namespace Microsoft.MaciOS.AssemblyComparator {
+	public class ModuleVisitor {
+		ModuleDefinition Module;
+		bool PublicOnly;
+		public ModuleVisitor (ModuleDefinition module, bool publicOnly)
+		{
+			Module = module;
+			PublicOnly = publicOnly;
+		}
+
+		public ModuleVisitor (Stream stm, bool publicOnly)
+			: this (ModuleDefinition.ReadModule (stm), publicOnly)
+		{
+		}
+
+
+		public void Visit ()
+		{
+			VisitTypes (Module.Types);
+		}
+
+		void VisitTypes (IEnumerable<TypeDefinition> types)
+		{
+			foreach (var type in types) {
+				VisitType (type);
+			}
+		}
+
+		void VisitType (TypeDefinition type)
+		{
+			if (PublicOnly && !type.IsPublic)
+				return;
+			TypeVisited.Invoke (this, new TypeVisitedEventArgs (VisitKind.Start, type));
+
+			if (type.NestedTypes.Count > 0)
+				VisitTypes (type.NestedTypes);
+			VisitFields (type);
+			VisitEvents (type);
+			VisitProperties (type);
+			VisitMethods (type);
+
+			TypeVisited.Invoke (this, new TypeVisitedEventArgs (VisitKind.End, type));
+		}
+
+		void VisitFields (TypeDefinition type)
+		{
+			foreach (var field in type.Fields)
+				VisitField (type, field);
+		}
+
+		void VisitField (TypeDefinition type, FieldDefinition field)
+		{
+			if (PublicOnly && !field.IsPublic)
+				return;
+
+			FieldVisited.Invoke (this, new MemberVisitedEventArgs<FieldDefinition> (VisitKind.Start, type, field));
+			FieldVisited.Invoke (this, new MemberVisitedEventArgs<FieldDefinition> (VisitKind.End, type, field));
+		}
+
+		void VisitProperties (TypeDefinition type)
+		{
+			foreach (var prop in type.Properties) {
+				VisitProperty (type, prop);
+			}
+		}
+
+		void VisitProperty (TypeDefinition type, PropertyDefinition prop)
+		{
+			if (PublicOnly && !(PublicGetterExists (prop) || PublicSetterExists (prop)))
+				return;
+			PropertyVisited (this, new MemberVisitedEventArgs<PropertyDefinition> (VisitKind.Start, type, prop));
+			PropertyVisited (this, new MemberVisitedEventArgs<PropertyDefinition> (VisitKind.End, type, prop));
+		}
+
+		bool PublicGetterExists (PropertyDefinition prop)
```

Comment:
This is a style suggestion. I find these helper methods really make the code more readable, but 4 lines make them pretty heavyweight to read. A one liner like this:

```suggestion
		bool PublicGetterExists (PropertyDefinition prop) => MethodExistsAndIsPublic (prop.GetMethod);
```

Makes them much easier to read IMO.